tests/ui/consts/static-mut-refs.rs

```diff
@@ -17,7 +17,9 @@ static mut INT_RAW: *mut isize = &mut 1isize as *mut _;
 pub fn main() {
     unsafe {
         TEST[0] += 1;
+        //~^ WARN creating a shared reference to mutable static is discouraged [static_mut_refs]
```

Comment:
I'm sure there's a good reason for it, but it's surprising that it warns about creating a shared reference rather than a mutable reference here.

Reply:
Good catch! Let's fix that before merging.

Reply:
Fixing this is hard and complex because we should handle the whole statement instead of the index expression. If we emit an error for the whole statement, another repeated error will be emitted for the index expression.

Can we emit a generic version in this case something like: `creating a reference to mutable static is discouraged`.

@michaelwoerister

Reply:
Or having such a hash map to store `hir_id` for index expressions when handling the whole statement to avoid handling the index expression twice. However, the question is where to store the hash map.

Reply:
> Can we emit a generic version in this case something like

That seems like a good solution. How about this if mutability is unknown:
> * this reference has lifetime `'static`, but if the static gets accessed in any way incompatible with this reference's mutability, or any other reference with incompatible mutability is created, then any further use of this reference is Undefined Behavior

Reply:
How about:
`This reference has lifetime `'static`. If it is accessed (read or written) by any other means, or any other reference is created, then any further use of this reference is Undefined Behavior`.

Reply:
That is only partially true. If this is a shared reference, then it is OK to read from the static or create other shared references.

Reply:
any suggestion?

Reply:
I'm kind of partial to the original simple proposal:

> creating a reference to a mutable static is discouraged

...because that is what we're discouraging, and it's so easy to get this slightly wrong, so we should say something simple as the main warning, and then perhaps add additional help notes and whatnot to supplement it.

For those notes, I'd adapt the language from @RalfJung [here](https://github.com/rust-lang/rust/pull/117556#issuecomment-1813053728):

> NOTE shared references to mutable statics are dangerous since if there's any kind of mutation of that static while the reference lives, that's undefined behavior; use `addr_of!` instead to create a raw pointer
>
> NOTE mutable references to mutable statics are dangerous since if there's any other pointer or reference used for that static while the reference lives, that's undefined behavior; use `addr_of_mut!` instead to create a raw pointer

As RalfJ points out, these really are two different cases, so I'd suggest just keeping them separate and printing out both notes if we're not sure which one is relevant.

Reply:
That sounds good: update the notes to not refer to "this reference". If mutability is known, give only the relevant note. If mutability is not known, give both notes.